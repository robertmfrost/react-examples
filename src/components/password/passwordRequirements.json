{
  "requirements": [
    {
      "rule": ".{6,}",
      "message": "Password must have at least six characters"
    },
    {
      "rule": "[A-Z]",
      "message": "Password must have at least one uppercase character"
    },
    {
      "rule": "[a-z]",
      "message": "Password must have at least one lowercase character"
    },
    {
      "rule": "[0-9]",
      "message": "Password must have at least one number"
    },
    {
      "rule": "[!@#$%^&*()_\\-+={[}\\]|:;\"'<,>.]",
      "message": "Password must have at least one special character: !@#$%^&*()_-+={[}]|:;\"'<,>."
    }
  ],
  "match": "Passwords must match",
  "changed": "Your password has been changed"
}