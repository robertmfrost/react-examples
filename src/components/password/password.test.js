import { render, screen, fireEvent } from '@testing-library/react';
import Password from './';
import { changed, match, requirements } from './passwordRequirements.json';

test('matches snapshot', () => {
  const { asFragment } = render(<Password />);
  expect(asFragment()).toMatchSnapshot();
});

test('form submit produces just min-character error when only that check fails', () => {
  render(<Password />);
  screen.getByTestId('password-new-input').value = 'Wrd1!';
  screen.getByTestId('password-retype-input').value = 'Wrd1!';
  fireEvent.submit(screen.getByTestId('password-form'));
  requirements.filter(r => r.rule !== '.{6,}').forEach(r => {
    expect(screen.queryByText(r.message)).toBeNull();
  });
  expect(screen.getByText(requirements.find(r => r.rule === '.{6,}').message)).toBeInTheDocument();
});

test('form submit produces just uppercase error when only that check fails', () => {
  render(<Password />);
  screen.getByTestId('password-new-input').value = 'password1!';
  screen.getByTestId('password-retype-input').value = 'password1!';
  fireEvent.submit(screen.getByTestId('password-form'));
  requirements.filter(r => r.rule !== '[A-Z]').forEach(r => {
    expect(screen.queryByText(r.message)).toBeNull();
  });
  expect(screen.getByText(requirements.find(r => r.rule === '[A-Z]').message)).toBeInTheDocument();
});

test('form submit produces just lowercase error when only that check fails', () => {
  render(<Password />);
  screen.getByTestId('password-new-input').value = 'PASSWORD1!';
  screen.getByTestId('password-retype-input').value = 'PASSWORD1!';
  fireEvent.submit(screen.getByTestId('password-form'));
  requirements.filter(r => r.rule !== '[a-z]').forEach(r => {
    expect(screen.queryByText(r.message)).toBeNull();
  });
  expect(screen.getByText(requirements.find(r => r.rule === '[a-z]').message)).toBeInTheDocument();
});

test('form submit produces just number error when only that check fails', () => {
  render(<Password />);
  screen.getByTestId('password-new-input').value = 'Password!';
  screen.getByTestId('password-retype-input').value = 'Password!';
  fireEvent.submit(screen.getByTestId('password-form'));
  requirements.filter(r => r.rule !== '[0-9]').forEach(r => {
    expect(screen.queryByText(r.message)).toBeNull();
  });
  expect(screen.getByText(requirements.find(r => r.rule === '[0-9]').message)).toBeInTheDocument();
});

test('form submit produces just special character error when only that check fails', () => {
  render(<Password />);
  screen.getByTestId('password-new-input').value = 'Password1';
  screen.getByTestId('password-retype-input').value = 'Password1';
  fireEvent.submit(screen.getByTestId('password-form'));
  requirements.filter(r => r.rule !== '[!@#$%^&*()_\\-+={[}\\]|:;\"\'<,>.]').forEach(r => {
    expect(screen.queryByText(r.message)).toBeNull();
  });
  expect(screen.getByText(requirements.find(r => r.rule === '[!@#$%^&*()_\\-+={[}\\]|:;\"\'<,>.]').message)).toBeInTheDocument();
});

test('form submit produces all errors when no checks pass', () => {
  render(<Password />);
  screen.getByTestId('password-new-input').value = ' ';
  screen.getByTestId('password-retype-input').value = '';
  fireEvent.submit(screen.getByTestId('password-form'));
  requirements.forEach(r => {
    expect(screen.getByText(r.message)).toBeInTheDocument();
  });
  expect(screen.getByText(match)).toBeInTheDocument();
});

test('form submit produces success message when all checks pass', () => {
  render(<Password />);
  screen.getByTestId('password-new-input').value = 'Password1!';
  screen.getByTestId('password-retype-input').value = 'Password1!';
  fireEvent.submit(screen.getByTestId('password-form'));
  expect(screen.getByText(changed)).toBeInTheDocument();
});
