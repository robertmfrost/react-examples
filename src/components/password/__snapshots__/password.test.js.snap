// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="password-page"
  >
    <form
      class="password-form"
      data-testid="password-form"
    >
      <input
        class="password-input"
        data-testid="password-new-input"
        placeholder="New Password"
        type="password"
      />
      <input
        class="password-input"
        data-testid="password-retype-input"
        placeholder="Retype Password"
        type="password"
      />
      <input
        class="password-submit"
        data-testid="password-submit"
        type="submit"
        value="Change Password"
      />
    </form>
  </main>
</DocumentFragment>
`;
